### ISSUE TITLE
Is it practical to have the GET version altogether?

### ISSUE THREAD

**Issue by chrysn** [core-wg/draft-dns-over-coap#2]
All I've found so far on why there's both the GET and the POST version of DoH in HTTP is that GETs are cachable.

In CoAP, we have [FETCH](https://datatracker.ietf.org/doc/html/rfc8132), which is just there for cases like this, and does away with both contender's downsides:
* Compared to POST, it's cachable (and safe and thus idempotent)
* Compared to GET, it doesn't need
  * URI template handling (which I'd discourage in an embedded context -- it's reasonably simple here, but I wouldn't want to have to think of the different rules that'd apply if the template occurs in the path or query string), and
  * base64 encoding (which adds both processing and wire overhead).

The only upside I see of keeping GET/POST over FETCH is that DoC could then be implemented by just cross-proxying to a DoH server, but given that such a proxy is unconstrained, reasonably straightforward to implement on its own and can thus take load off the constrained client, I advocate going all FETCH.

**Comment by miri64** [comment]
Another drawback of GET (as [pointed out in the current draft version](https://github.com/anr-bmbf-pivot/draft-dns-over-coaps/blob/f5950cf75af68c6a4addec55942b76573aac27ec/draft-lenders-dns-over-coaps.md#dns-queries-in-coap-requests)) is that with it block-wise transfer is not possible with large requests.

**Comment by miri64** [comment]
Ok, I think we agree for now that in the long run, we should remove GET and POST. However, at least for version `-00` of the draft I would like to keep it in for two reasons: a) to document our thought process for posterity and b) in case we find a use case that requires GET and need to revert back to the current version of CoAP messaging. I documented the removal of GET and POST by linking this issue in the future Change Log section in https://github.com/anr-bmbf-pivot/draft-dns-over-coaps/commit/a8dcfd4ab3087682aa2191169ff2cb93554b4616.

**Comment by waehlisch** [comment]
Do we have an overview of `FETCH` support in CoAP implementations?

**Comment by chrysn** [comment]
I don't; Francesca and I started [preparing a survey](https://github.com/core-wg/wiki/wiki/Implementations-overview), but didn't even finish the requirements.

**Comment by waehlisch** [comment]
oops, ... then we do a brief survey offline. we need an overview, which implementation supports `FETCH`, for example.

**Comment by chrysn** [comment]
If it's a helpful data point: In all the interops and implementation discussions around OSCORE that I can remember (and I've been in many of them), never has any implementer complained about the use of FETCH that'd have kept them from using OSCORE with that library.

**Comment by miri64** [comment]
> oops, ... then we do a brief survey offline. we need an overview, which implementation supports FETCH, for example.

Is that really necessary? As [mentioned above](https://github.com/anr-bmbf-pivot/draft-dns-over-coaps/issues/2#issuecomment-895136457), adding `FETCH` support is just a matter of adding a FETCH = 5 constant and using it in an application in accordance with RFC 8132. Check all implementations in [the list](http://coap.technology/impls.html) if they do that, on the other hand, seems like much more work.

**Comment by miri64** [comment]
The only nodes where FETCH is interesting is at proxies that cache. Here an (equally minor) adaptation needs to be made, that the payload of a FETCH request is considered for the cache key.

**Comment by waehlisch** [comment]
hard to follow. will talk offline.

**Comment by miri64** [comment]
> Is that really necessary? As [mentioned above](https://github.com/anr-bmbf-pivot/draft-dns-over-coaps/issues/2#issuecomment-895136457), adding `FETCH` support is just a matter of adding a FETCH = 5 constant and using it in an application in accordance with RFC 8132. Checking all implementations in [the list](http://coap.technology/impls.html) if they do that, on the other hand, seems like much more work.

Regardless, I looked a bit into this, and the data is a bit crushing. Of the 19 implementations I looked into so far, only 6 support FETCH. Granted, most of these implementations haven't received an update in years (many last before the publication of RFC8132), but even some implementations I would call reference implementations such as Erbium (in all its flavors known to me, Contiki, Contiki-NG, and Wakaama) do not support FETCH. :confused:
